### PR TITLE
[Do NOT pull] Implement LSTM

### PR DESCRIPTION
**Pull request by rihardsk**
This is meant to share the progress on the LSTM implementation in Marian. See #78. Any comments are welcome.

### CODE REVIEW

**Comment by tomekd** [comment]
Hi @rihardsk 
good job so far. 
One comment on coding-style: never use ``use namespace``, especially in headers.

Are you planning to implement the CPU version too?

Best

**Comment by rihardsk** [comment]
@emjotde ok, thanks for pointing that out. As a heads up - there might be more such issues as I don't use C++ that often.

Yes, I don't expect it to be too much work after the GPU version is finished. I guess there shouldn't be too much differences in the implementations?

**Comment by hieuhoang** [comment]
can you pls check in gpu/dl4mt/cell.h

**Comment by hieuhoang** [comment]
closing this for the time being. Doesn't compile so shouldn't be pulled

**Comment by rihardsk** [comment]
@hieuhoang Well this is still work in progress as indicated by the `[WIP]` prefix and as hinted in the first post. I opened this pull request to have a place for discussion while I'm implementing the feature.

**Comment by hieuhoang** [comment]
ok. Just making clear no-one should execute this pull request. cell.h would be good

**Comment by rihardsk** [comment]
@hieuhoang  I have pushed more work to [tilde-nlp/amunmt/commits/lstm](]https://github.com/tilde-nlp/amunmt/commits/lstm) but the changes are no longer broadcast here since the PR is closed.

IMHO a work-in-progress PR is the most suitable place for a discussion on the changes so it would be great if the PR was re-opened. I have included the missing `cell.*` files so the code is in a working state now.

**Comment by hieuhoang** [comment]
sure thing. Changed the title so there's no mistaking that you don't want it pulled

**Comment by rihardsk** [comment]
Thanks! Ok then, a bit more on the changes - I've refactored the code so that the cell's state is encapsulated in [`CellState`](https://github.com/tilde-nlp/amunmt/blob/29f8316a127420b3e89b53dad39f9a8eda553727/src/amun/gpu/dl4mt/cellstate.h#L9) with `CellState.cell` for the cell's internal state, and `CellState.output` for the cell's output state.

I opted to not use a `nullptr` value to indicate that `CellState.cell` is not used (which is the case for GRU). Instead I leave the `CellState.cell` matrix empty (having `CellState.cell->size() == 0`). The reason for it is that using a `nullptr` complicates the initialization of `CellState` a bit - I'd have to figure out how to make [`EncoderDecoder::NewState()`](https://github.com/tilde-nlp/amunmt/blob/29f8316a127420b3e89b53dad39f9a8eda553727/src/amun/gpu/decoder/encoder_decoder.cu#L51) initialize different values for different Encoder and Decoder cell types.

**Comment by rihardsk** [comment]
I have implemented the slow version of LSTM, but before pushing, I need to test it on a trained model. So I used Marian to train one but ended up with something incompatible with Amun. Here are the keys from the model trained with Marian:
```
['decoder_U_att',
 'decoder_W_comb_att',
 'decoder_Wc_att',
 'decoder_Wemb',
 'decoder_att_gamma1',
 'decoder_att_gamma2',
 'decoder_b_att',
 'decoder_cell1_U',
 'decoder_cell1_W',
 'decoder_cell1_b',
 'decoder_cell1_gamma1',
 'decoder_cell1_gamma2',
 'decoder_cell2_U',
 'decoder_cell2_W',
 'decoder_cell2_b',
 'decoder_cell2_gamma1',
 'decoder_cell2_gamma2',
 'decoder_ff_logit_l1_W0',
 'decoder_ff_logit_l1_W1',
 'decoder_ff_logit_l1_W2',
 'decoder_ff_logit_l1_b0',
 'decoder_ff_logit_l1_b1',
 'decoder_ff_logit_l1_b2',
 'decoder_ff_logit_l1_gamma0',
 'decoder_ff_logit_l1_gamma1',
 'decoder_ff_logit_l1_gamma2',
 'decoder_ff_logit_l2_W',
 'decoder_ff_logit_l2_b',
 'decoder_ff_state_W',
 'decoder_ff_state_b',
 'decoder_ff_state_gamma',
 'encoder_Wemb',
 'encoder_bi_U',
 'encoder_bi_W',
 'encoder_bi_b',
 'encoder_bi_gamma1',
 'encoder_bi_gamma2',
 'encoder_bi_r_U',
 'encoder_bi_r_W',
 'encoder_bi_r_b',
 'encoder_bi_r_gamma1',
 'encoder_bi_r_gamma2',
 'special:model.yml']
```
And here are the keys of an older model trained with Nematus that loads fine in Amun:
```
['ff_logit_W',
 'Wemb',
 'encoder_U',
 'encoder_W',
 'encoder_r_Wx',
 'decoder_U',
 'decoder_W',
 'encoder_bx',
 'ff_logit_lstm_W',
 'ff_logit_prev_W',
 'ff_state_W',
 'decoder_b',
 'decoder_Wc',
 'ff_logit_prev_b',
 'ff_logit_lstm_b',
 'decoder_b_att',
 'decoder_bx_nl',
 'decoder_Wcx',
 'ff_logit_b',
 'decoder_Ux',
 'ff_state_b',
 'encoder_b',
 'decoder_bx',
 'encoder_r_bx',
 'decoder_Wc_att',
 'ff_logit_ctx_b',
 'ff_logit_ctx_W',
 'decoder_U_att',
 'encoder_r_U',
 'encoder_r_b',
 'decoder_c_tt',
 'Wemb_dec',
 'decoder_U_nl',
 'encoder_r_W',
 'decoder_W_comb_att',
 'decoder_b_nl',
 'decoder_Wx',
 'uidx',
 'decoder_Ux_nl',
 'encoder_Ux',
 'history_errs',
 'encoder_Wx',
 'encoder_r_Ux']
```
Why does Marian suddenly use a different naming pattern for the model keys?

**Comment by emjotde** [comment]
If I am not wrong, the LSTM cell is currently working with the `s2s` model type which has these different matrix names. You would need to get the `amun` model type in Marian to understand the LSTM cells. Or I can take a look at that after WMT. 

In the `s2s` model I am not keeping the Nematus naming convention as I do not like it particularly well. The `amun` model type is a converter on top of `s2s`.

**Comment by rihardsk** [comment]
The LSTM implementation contains a bug which makes it crash somewhere in `NthElement::GetPairs()`. I'm pushing it anyway because I've spent too much time trying to hunt it down already. If anybody is willing to assist in debugging this I can provide you with a Amun style LSTM model to run it with.

(The MLSTM implementation probably has a bug of it's own but I'm not concerned with it currently.)

**Comment by hieuhoang** [comment]
yes please, model would be good

Hieu Hoang
http://moses-smt.org/

**Comment by rihardsk** [comment]
@hieuhoang I've sent you an email with a link to the model.

**Comment by hieuhoang** [comment]
it appears to be this line in lstm.h
         Element(_1 * _2, Temp2_, F_);
the 2 matrices have different sizes on the 2nd decoding iteration
   Temp2_=1x1024x1x1
   F_=11x1024x1x1

I'm not sure why it's happening.

I've merged with master and added debugging message to the code, and pushed it out to the main reposit, lstm branch
   https://github.com/marian-nmt/marian/tree/lstm
if you have more time, you can take a look at it

**Comment by rihardsk** [comment]
@hieuhoang thanks, I was debugging in all the wrong places due to not knowing what `cudaStreamSynchronize` is.

For me the size of the `F_` matrix is fine. I always get `F_=1x1024x1x1=1024`. But where it fails for me is:
```c++
Slice(O_, FIO_, 2, cols);
```
Which happens because `FIO_`'s shape suddenly becomes `FIO_=1x2048x1x1=2048`  instead of `FIO_=1x3072x1x1=3072` as it should be.

I'm looking into why this is happening. Maybe this is somehow related to what you're seeing.

**Comment by hieuhoang** [comment]
cool. the code is looking good.

I also though bugs were in nth_element for a long time and had to learn to use cudaStreamSynchronize  debugging. Think of GPU programming as controlling 2 independent computers. They are only guaranteed to communicate at the sync point where they will tell each other if something is wrong

The dimension mismatch - it may be that you are confusing the batch size (1) with the number of hypotheses (between 1 to batch-size*beam-size). 

In the 1st iteration, the number of hypothees is always batch-size (or less if the test set is small or it's at the end of the test set). In the 2nd and subsequent iterations it can be up to (batch-size*beam-size)

**Comment by rihardsk** [comment]
The dimension mismatch was caused by GRU not passing along `State.cell` which it receives from a previous LSTM cell (I used GRU as the second cell in the decoder, the bug wouldn't have been triggered if I had a model with both LSTM cells). Which in turn caused [this _if_](https://github.com/tilde-nlp/amunmt/blob/f142137dbe2efd88c2f4d4d4cd87d65146cff045/src/amun/gpu/decoder/encoder_decoder.cu#L136) to not get triggered which interfered with assembling the beam state.

SlowLSTM should be working now, but I don't have a real Amun-compatible LSTM model to test it on. I intend to verify it with an MLSTM model after I've finished debugging the multiplicative cell wrapper.

@emjotde: to verify - you aren't doing the _forget gate bias trick_ in Marian, right? ([Here's a description of it in section 2.2](http://proceedings.mlr.press/v37/jozefowicz15.pdf) - a bias is added so that the forget gate doesn't discard everything at the beginning of training.) We do this internally but for now I'll just match anything that's in Marian.

**Comment by rihardsk** [comment]
The dimension mismatch was caused by GRU not passing along `State.cell` which it receives from a previous LSTM cell (I used GRU as the second cell in the decoder, the bug wouldn't have been triggered if I had a model with both LSTM cells). Which in turn caused [this _if_](https://github.com/tilde-nlp/amunmt/blob/f142137dbe2efd88c2f4d4d4cd87d65146cff045/src/amun/gpu/decoder/encoder_decoder.cu#L136) to not get triggered which interfered with assembling the beam state.

SlowLSTM should be working now, but I don't have a real Amun-compatible LSTM model to test it on. I intend to verify it with an MLSTM model after I've finished debugging the multiplicative cell wrapper.

@emjotde: to verify - you aren't doing the _forget gate bias trick_ in Marian, right? ([Here's a description of it in section 2.2](http://proceedings.mlr.press/v37/jozefowicz15.pdf) - a bias is added so that the forget gate doesn't discard everything at the beginning of training.) We do this internally but for now I'll just match anything that's in Marian.

**Comment by emjotde** [comment]
You would not need to match this in Amun anywhere, this is only an initialization setting for training. 

However, I agree it would probably make a lot of sense to set that to a similar value by default. I have done similar things for highway connections where intializing the bias a reasonably large negative value means "carry over" when training starts. It's easy enough to change in Marian. Will try it out.

**Comment by rihardsk** [comment]
SlowLSTM is done, I think. Running Tilde's MLSTM models produces the same outputs as with our Nematus implementation.

Next I'll incorporate the CUDA kernel from Marian to know for sure that the implementations match. But I think this branch could be merged before that. @hieuhoang what do you think?

@emjotde you're right, hadn't realized it earlier. Anyway, the LSTM implementation isn't affected in any way, I just converted our MLSTM models.

**Comment by hieuhoang** [comment]
excellent. Yep, lets merge, unless anyone here objects. Make sure you run regression tests
   ln -s marian.lstm marian
   git clone https://github.com/marian-nmt/marian-regression-tests.git
   cd marian-regression-tests
   make download_models
   ./run_tests.sh
It may fail on 1 or 2 tests due to numerical differences, ignore it. If there are big differences or segfaults, then u should worry

Since this is your 1st time & its a big checkin, i'll run the test too when I get home

**Comment by hieuhoang** [comment]
do you have a model for lstm we can use to test your changes? We can make it into a regression test so that the model will be protected into the future

**Comment by emjotde** [comment]
I understand this is currently only SlowLSTM specific?

**Comment by hieuhoang** [comment]
all tests passed. Merged. @rihardsk  - i'm giving you commit access so you can tweak and fix bugs yourself.

Please follow these house rules when you commit to master:
  1. if you commit, please listen to the mailing list/forum for a few weeks in case someone reports an issue with your commit
  2. don't commit large data files or binary files. Some people have slow network connections and checking them out can be a pain.
  3. please run the regression test to make sure you haven't broken anything people care about
  4. if you pull someone else's changes, the above also applies.

**Comment by rihardsk** [comment]
Ran the tests, they pass.

I don't have a pure LSTM model. I have only ones with MLSTM cells. We could use the English-Romanian model that I sent you for debugging, but I'd have to convert it first (some matrices must to be merged/renamed).

To add regression tests for an LSTM model I could train something new using Marian. Or perhaps @emjotde already has something suitable? In any case the model key names would have to be converted to match the Amun format.

@emjotde the implementation currently only contains SlowLSTM, if that's what you mean?

**Comment by rihardsk** [comment]
@hieuhoang have you given me the commit rights yet? I can't push to the master. I want to make some small changes for the multiplicative wrapper matrix naming.

**Comment by hieuhoang** [comment]
i've sent you an invite, accept it on github & you're good to go

**Comment by hieuhoang** [comment]
Do you have a mlstm that does something interesting? The 1 you gave me last month doesn't give any output, but at least it doesn't crash.

How are you training these mlstm models? Your own fork of marian?

**Comment by rihardsk** [comment]
Yup, it's with our own fork. It hasn't been upstreamed because refactoring Nematus to support mlstm alongside gru was simply too much work at the time of implementing it.

The model I sent you needs to be converted. I'll do it some time later today or tomorrow, and send it to you.

Btw, how should we go about supporting Marian-trained s2s models? It would be nice to be able to run lstm/mlstm models trained with Marian. I'm thinking of creating a converter tool that would rename the matrices, bet currently I don't have time to figure out Marian matrix naming.

**Comment by hieuhoang** [comment]
cool. In your own time, no hurry

**Comment by hieuhoang** [comment]
I don't know what to do with general marian s2s models. It may be that amun will implement specific models where you need to speed, or particular features that are difficult to put into s2s. For the rest, s2s should be able to run the model

**Comment by emjotde** [comment]
As I said earlier, my long term plan is to retire amun as soon as the marian engine can emulate amun 1-1. We are getting a lot closer to that. I will work on batched translation in the coming days and then the only thing left is a CPU backend (are you guys using that?).

Once we are there I was thinking of creating an Amun binary with the same inferface as the original Amun with Marian behind it as a drop-in replacement. This binary would automatically allow to use anything that is implemented in Marian. Not sure how much sense it makes to do things by hand for Amun until then.

**Comment by hieuhoang** [comment]
yes, batching would be good.

Everyone's got plans. I'll still be working on amun in the forseeable future on things that don't depend on the latest models, eg. FPGA, half precision, faster beams etc

**Comment by rihardsk** [comment]
I checked the en-ro mlstm model I gave you. I wasn't able to get it to work properly. Something must have gone wrong during training. I'll retrain it sometime soon.

**Comment by hieuhoang** [comment]
@rihardsk  Do you have a model I can create a regression test with? I'm about to merge changes into master that will make it 30% faster. I wanna make sure the mLSTM isn't broken when I do

**Comment by rihardsk** [comment]
No, I don't have on yet, sorry. We had other things to train so there weren't any free gpu resources to use. If you're willing to wait a few days, I can start training one now. I'll use the EN-RO WMT dataset.

**Comment by hieuhoang** [comment]
sure, i few days is fine

**Comment by emjotde** [comment]
@hieuhoang can you elaborate what you did and if some of this would be usable in marian for speeding up training?

**Comment by hieuhoang** [comment]
[10-robertlim.pdf](https://github.com/marian-nmt/marian/files/1474789/10-robertlim.pdf)

Basically, combining the softmax and nth_element into 1 function. Can help training - only needs 1-best so code can be even simplier. Can also help s2s

**Comment by emjotde** [comment]
This gives a 30% total speed improvement?

**Comment by hieuhoang** [comment]
yep, as measured on my 940mx. On the 1080 its about 20%

**Comment by emjotde** [comment]
Interesting, when profiling I never saw the softmax + nth_element to take that much time that you can actually save 20-30% by improving that.

**Comment by hieuhoang** [comment]
Whats your model, batch, beam size?

Hieu Hoang
Sent while bumping into things

**Comment by hieuhoang** [comment]
And vocab size

Hieu Hoang
Sent while bumping into things

**Comment by emjotde** [comment]
vocab 32,000; beam 6-12.

**Comment by tomekd** [comment]
Sounds good!

**Comment by rihardsk** [comment]
@hieuhoang I've sent you the mlstm model in email.

**Comment by hieuhoang** [comment]
cheers. Added you models to regression test and merged my changes
